from __future__ import unicode_literals

import logging
import sys
import warnings
from functools import wraps

from django.conf import settings
from django.core import signals
from django.core.exceptions import PermissionDenied, SuspiciousOperation
from django.http import Http404
from django.http.multipartparser import MultiPartParserError
from django.urls import get_resolver, get_urlconf
from django.utils import six
from django.utils.decorators import available_attrs
from django.utils.deprecation import RemovedInDjango20Warning
from django.utils.encoding import force_text
from django.views import debug

logger = logging.getLogger('django.request'***REMOVED***


def convert_exception_to_response(get_response***REMOVED***:
    ***REMOVED***
    Wrap the given get_response callable in exception-to-response conversion.

    All exceptions will be converted. All known 4xx exceptions (Http404,
    PermissionDenied, MultiPartParserError, SuspiciousOperation***REMOVED*** will be
    converted to the appropriate response, and all other exceptions will be
    converted to 500 responses.

    This decorator is automatically applied to all middleware to ensure that
    no middleware leaks an exception and that the next middleware in the stack
    can rely on getting a response instead of an exception.
    ***REMOVED***
    @wraps(get_response, assigned=available_attrs(get_response***REMOVED******REMOVED***
    def inner(request***REMOVED***:
        ***REMOVED***
            response = get_response(request***REMOVED***
        except Exception as exc:
            response = response_for_exception(request, exc***REMOVED***
        return response
    return inner


def response_for_exception(request, exc***REMOVED***:
    if isinstance(exc, Http404***REMOVED***:
        if settings.DEBUG:
            response = debug.technical_404_response(request, exc***REMOVED***
        else:
            response = get_exception_response(request, get_resolver(get_urlconf(***REMOVED******REMOVED***, 404, exc***REMOVED***

    elif isinstance(exc, PermissionDenied***REMOVED***:
        logger.warning(
            'Forbidden (Permission denied***REMOVED***: %s', request.path,
            extra={'status_code': 403, 'request': request***REMOVED***,
        ***REMOVED***
        response = get_exception_response(request, get_resolver(get_urlconf(***REMOVED******REMOVED***, 403, exc***REMOVED***

    elif isinstance(exc, MultiPartParserError***REMOVED***:
        logger.warning(
            'Bad request (Unable to parse request body***REMOVED***: %s', request.path,
            extra={'status_code': 400, 'request': request***REMOVED***,
        ***REMOVED***
        response = get_exception_response(request, get_resolver(get_urlconf(***REMOVED******REMOVED***, 400, exc***REMOVED***

    elif isinstance(exc, SuspiciousOperation***REMOVED***:
        # The request logger receives events for any problematic request
        # The security logger receives events for all SuspiciousOperations
        security_logger = logging.getLogger('django.security.%s' % exc.__class__.__name__***REMOVED***
        security_logger.error(
            force_text(exc***REMOVED***,
            extra={'status_code': 400, 'request': request***REMOVED***,
        ***REMOVED***
        if settings.DEBUG:
            response = debug.technical_500_response(request, *sys.exc_info(***REMOVED***, status_code=400***REMOVED***
        else:
            response = get_exception_response(request, get_resolver(get_urlconf(***REMOVED******REMOVED***, 400, exc***REMOVED***

    elif isinstance(exc, SystemExit***REMOVED***:
        # Allow sys.exit(***REMOVED*** to actually exit. See tickets #1023 and #4701
        raise

    else:
        signals.got_request_exception.send(sender=None, request=request***REMOVED***
        response = handle_uncaught_exception(request, get_resolver(get_urlconf(***REMOVED******REMOVED***, sys.exc_info(***REMOVED******REMOVED***

    return response


def get_exception_response(request, resolver, status_code, exception, sender=None***REMOVED***:
    ***REMOVED***
        callback, param_dict = resolver.resolve_error_handler(status_code***REMOVED***
        # Unfortunately, inspect.getargspec result is not trustable enough
        # depending on the callback wrapping in decorators (frequent for handlers***REMOVED***.
        # Falling back on try/***REMOVED***
        ***REMOVED***
            response = callback(request, **dict(param_dict, exception=exception***REMOVED******REMOVED***
        except TypeError:
            warnings.warn(
                "Error handlers should accept an exception parameter. Update "
                "your code as this parameter will be required in Django 2.0",
                RemovedInDjango20Warning, stacklevel=2
            ***REMOVED***
            response = callback(request, **param_dict***REMOVED***
    except Exception:
        signals.got_request_exception.send(sender=sender, request=request***REMOVED***
        response = handle_uncaught_exception(request, resolver, sys.exc_info(***REMOVED******REMOVED***

    return response


def handle_uncaught_exception(request, resolver, exc_info***REMOVED***:
    ***REMOVED***
    Processing for any otherwise uncaught exceptions (those that will
    generate HTTP 500 responses***REMOVED***.
    ***REMOVED***
    if settings.DEBUG_PROPAGATE_EXCEPTIONS:
        raise

    logger.error(
        'Internal Server Error: %s', request.path,
        exc_info=exc_info,
        extra={'status_code': 500, 'request': request***REMOVED***,
    ***REMOVED***

    if settings.DEBUG:
        return debug.technical_500_response(request, *exc_info***REMOVED***

    # If Http500 handler is not installed, reraise the last exception.
    if resolver.urlconf_module is None:
        six.reraise(*exc_info***REMOVED***
    # Return an HttpResponse that displays a friendly error message.
    callback, param_dict = resolver.resolve_error_handler(500***REMOVED***
    return callback(request, **param_dict***REMOVED***
